S3-plain based disk supporting directory rename

jkartseva opened the pull request:
Fixes: https://github.com/ClickHouse/ClickHouse/issues/58347

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a rewritable S3 disk which supports INSERT operations and does not require locally stored metadata.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

jkartseva committed:
S3-plain based disk supporting directory rename

Fixes: https://github.com/ClickHouse/ClickHouse/issues/58347

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #61116 from jkartseva/ClickHouse:s3-plain-with-replace -->
*This is an automated comment for commit dee30c2eb069f33536a2e6c5fd5a2d92b27a7c16 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61116/dee30c2eb069f33536a2e6c5fd5a2d92b27a7c16/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

jkartseva committed:
Do not create directory metadata recursively

jkartseva committed:
concurrency control

jkartseva committed:
better transaction rollback

jkartseva committed:
more tests

jkartseva committed:
add integration test

jkartseva committed:
Do not list prefix.path in listDirectory

jkartseva committed:
extend system logs integration tests

jkartseva committed:
docs

jkartseva committed:
improvements, cleanups, comments

jkartseva commented:
an example of s3 directory structure: https://pastila.nl/?000db959/a9e05708623112738049177aca209439#G5jPgWHeHVknjx4Tkz3L0g==

alexey-milovidov commented on docs/en/engines/table-engines/mergetree-family/mergetree.md:
```diff
@@ -769,6 +769,7 @@ In addition to local block devices, ClickHouse supports these storage types:
 - [`web` for read-only from web](#web-storage)
 - [`cache` for local caching](/docs/en/operations/storing-data.md/#using-local-cache)
 - [`s3_plain` for backups to S3](/docs/en/operations/backup#backuprestore-using-an-s3-disk)
+- [`s3_plain_rewritable` for immutable, non-replicated tables in S3](/docs/en/operations/storing-data.md#s3-plain-rewritable-storage)
```
What is an unbundled configuration option?

(what type of metadata over the s3 object storage?)

kssenii commented on tests/queries/0_stateless/03008_s3_plain_rewritable.sql:
```diff
@@ -0,0 +1,22 @@
+    secret_access_key = clickhouse);
+
+insert into test_mt (*) values (1, 2), (2, 2), (3, 1), (4, 7), (5, 10), (6, 12);
+insert into test_mt (*) select number, number from numbers_mt(1000000);
```
It will also be good to check and validate content of `prefix.path` file, in integration test in will probably be more convenient.

kssenii replied:
Also the layout https://pastila.nl/?000db959/a9e05708623112738049177aca209439#G5jPgWHeHVknjx4Tkz3L0g==
where we make random only part directory: `./store/df3/df3bc07c-0967-43f7-adbc-9daaaa8d629d/mlkufoyckwdnhxfo/` suggests that moving parts between tables will not work, e.g. for example `ALTER TABLE MOVE PARTITION TO TABLE`.
As I see https://github.com/ClickHouse/ClickHouse/issues/58347 suggests "Let the disk store each directory as a pseudorandom static prefix in S3", e.g. in my impression it means each directory should have a randomized name, while now it is not each. Seems there is no reason not to do it for each directory name, WDYT?

kssenii replied:
Also may be it will be good to make a test where a table has a projection, because it will make files layout more complex (one more directory layer).

jkartseva replied:
Currently, the disk does not support `moveFile` required for `materialize projections`
Also, from the task description (projection is mutation):
> Let's also forbid ALTERs and mutations for the tables on this disk

Query `alter table test_mt materialize projection test_mt_projection`:

```
6. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: void std::__function::__policy_invoker<void (std::shared_ptr<DB::IMetadataTransaction>)>::__call_impl<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::moveFile(String const&, String const&)::$_0, void (std::shared_ptr<DB::IMetadataTransaction>)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::IMetadataTransaction>&&) @ 0x000000000f3a0761
7. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::(anonymous namespace)::PureMetadataObjectStorageOperation::execute(std::shared_ptr<DB::IMetadataTransaction>) @ 0x000000000f39fe8f
8. ./build/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:891: DB::DiskObjectStorageTransaction::commit() @ 0x000000000f39f89b
9. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::DiskObjectStorage::moveFile(String const&, String const&, bool) @ 0x000000000f3905e8
10. ./contrib/llvm-project/libcxx/include/string:1499: DB::MergeTreeMutationEntry::commit(unsigned long) @ 0x0000000010ff158d
11. ./contrib/llvm-project/libcxx/include/string:1499: DB::StorageMergeTree::startMutation(DB::MutationCommands const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fc86b2
12. ./build/./src/Storages/StorageMergeTree.cpp:0: DB::StorageMergeTree::mutate(DB::MutationCommands const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fcb627
13. ./contrib/llvm-project/libcxx/include/vector:543: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb36b50
14. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb3434c
15. ./build/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x00000000101db067
```
I can implement `moveFile` with the limitation that the filename does not change and see what happens.

jkartseva replied:
Not related to the PR, it looks like there is a bug in change setting that became visible if the disk is specified through `SETTINGS disk = disk(...)`in `CREATE TABLE`  and then the table is altered:
```
create table test_mt (a Int32, b Int64) engine = MergeTree() order by a
settings disk = disk(
    type = s3_plain_rewritable,
    endpoint = 'http://localhost:11111/test/test_mt/',
    access_key_id = clickhouse,
    secret_access_key = clickhouse);

alter table test_mt add column c Int64 after b;
```

```
2024.04.25 01:03:07.530107 [ 114749 ] {2a3cd2ac-1f4e-4e29-a720-a75ad7ca8e41} <Error> TCPHandler: Code: 170. DB::Exception: Bad get: has CustomType, requested String. (BAD_GET), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x00000000134b1232
1. ./build/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000afab639
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000063d0ecc
3. DB::Exception::Exception<std::basic_string_view<char, std::char_traits<char>>, DB::Field::Types::Which const&>(int, FormatStringHelperImpl<std::type_identity<std::basic_string_view<char, std::char_traits<char>>>::type, std::type_identity<DB::Field::Types::Which const&>::type>, std::basic_string_view<char, std::char_traits<char>>&&, DB::Field::Types::Which const&) @ 0x00000000063dbd2b
4. auto& DB::Field::safeGet<String const&>() @ 0x00000000066d96d1
5. ./contrib/llvm-project/libcxx/include/string:1499: DB::MergeTreeSettingsTraits::Accessor::instance()::$_0::operator()() const::'lambda127'(DB::MergeTreeSettingsTraits::Data&, DB::Field const&)::__invoke(DB::MergeTreeSettingsTraits::Data&, DB::Field const&) @ 0x0000000010e6f40c
6. ./src/Core/BaseSettings.h:224: DB::BaseSettings<DB::MergeTreeSettingsTraits>::set(std::basic_string_view<char, std::char_traits<char>>, DB::Field const&) @ 0x0000000010e792ba
7. ./src/Core/BaseSettings.h:319: DB::BaseSettings<DB::MergeTreeSettingsTraits>::applyChange(DB::SettingChange const&) @ 0x0000000010d180e8
8. ./contrib/llvm-project/libcxx/include/__iterator/wrap_iter.h:100: DB::MergeTreeData::checkColumnFilenamesForCollision(DB::StorageInMemoryMetadata const&, bool) const @ 0x0000000010cc05db
9. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:3245: DB::MergeTreeData::checkAlterIsPossible(DB::AlterCommands const&, std::shared_ptr<DB::Context const>) const @ 0x0000000010cba24f
10. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:0: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb366f6
11. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb3434c
```

The type of the "disk" setting is set as `String`, that's why it's a `BAD_GET` error.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeSettings.h#L171

I'm not sure why it is trying to change the disk setting during `ALTER TABLE` when it is not even specified.

It works fine `storage_policy` setting.

jkartseva replied:
`ALTER TABLE MOVE PARTITION TO TABLE` requires hard link creation

``` sql
create table test_mt_dst (a Int32, b Int64) engine = MergeTree() partition by intDiv(a, 10000) order by tuple(a, b)
settings storage_policy = 's3_plain_rewritable';
alter table test_mt_src move partition 0 to table test_mt_dst;
```

```
4. ./src/Disks/ObjectStorages/IMetadataStorage.h:151: DB::IMetadataTransaction::throwNotImplemented() @ 0x000000000f3e2af9
5. ? @ 0x000000000f3e1c86
6. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: void std::__function::__policy_invoker<void (std::shared_ptr<DB::IMetadataTransaction>)>::__call_impl<std::__function::__default_alloc_func<DB::DiskObjectStorageTransaction::createHardLink(String const&, String const&)::$_0, void (std::shared_ptr<DB::IMetadataTransaction>)>>(std::__function::__policy_storage const*, std::shared_ptr<DB::IMetadataTransaction>&&) @ 0x000000000f3980c1
7. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::(anonymous namespace)::PureMetadataObjectStorageOperation::execute(std::shared_ptr<DB::IMetadataTransaction>) @ 0x000000000f38efef
8. ./build/./src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:891: DB::DiskObjectStorageTransaction::commit() @ 0x000000000f38e9fb
9. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::DiskObjectStorage::createHardLink(String const&, String const&, bool) @ 0x000000000f38030f
10. ./build/./src/Storages/MergeTree/localBackup.cpp:0: DB::(anonymous namespace)::localBackupImpl(std::shared_ptr<DB::IDisk> const&, DB::IDiskTransaction*, String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, bool, unsigned long, std::optional<unsigned long>, bool, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>> const&) @ 0x0000000010fa0685
11. ./build/./src/Storages/MergeTree/localBackup.cpp:0: DB::localBackup(std::shared_ptr<DB::IDisk> const&, String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, bool, std::optional<unsigned long>, bool, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>> const&, std::shared_ptr<DB::IDiskTransaction>) @ 0x0000000010f9f8f8
12. ./contrib/llvm-project/libcxx/include/string:1499: DB::DataPartStorageOnDiskBase::freeze(String const&, String const&, DB::ReadSettings const&, DB::WriteSettings const&, std::function<void (std::shared_ptr<DB::IDisk> const&)>, DB::IDataPartStorage::ClonePartParams const&) const @ 0x0000000010bce704
13. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:7098: DB::MergeTreeData::cloneAndLoadDataPartOnSameDisk(std::shared_ptr<DB::IMergeTreeDataPart const> const&, String const&, DB::MergeTreePartInfo const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::IDataPartStorage::ClonePartParams const&, DB::ReadSettings const&, DB::WriteSettings const&) @ 0x0000000010ce1cd6
14. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::StorageMergeTree::movePartitionToTable(std::shared_ptr<DB::IStorage> const&, std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>) @ 0x0000000010fce793
15. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:701: DB::MergeTreeData::movePartitionToTable(DB::PartitionCommand const&, std::shared_ptr<DB::Context const>) @ 0x0000000010cc6851
16. ./build/./src/Storages/MergeTree/MergeTreeData.cpp:5163: DB::MergeTreeData::alterPartition(std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::vector<DB::PartitionCommand, std::allocator<DB::PartitionCommand>> const&, std::shared_ptr<DB::Context const>) @ 0x0000000010cc75b7
17. ./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:815: DB::InterpreterAlterQuery::executeToTable(DB::ASTAlterQuery const&) @ 0x000000000fb25283
18. ./build/./src/Interpreters/InterpreterAlterQuery.cpp:71: DB::InterpreterAlterQuery::execute() @ 0x000000000fb2290c
```

From the task: 
> Let's also forbid ALTERs and mutations for the tables on this disk, so we never do hard link operations.

jkartseva replied:
> Let the disk store each directory as a pseudorandom static prefix in S3", e.g. in my impression it means each directory should have a randomized name, while now it is not each. Seems there is no reason not to do it for each directory name, WDYT?

That was my initial implementation.
I changed it because `prefix.path` will be created for directories that are static (will never be be renamed or moved), such as

the top-level subdirectories `store`, `7d5` and `7d563340-6f73-463b-bc78-95baf347d19a` (table uuid) in the example `store/7d5/7d563340-6f73-463b-bc78-95baf347d19a/tmp_insert_all_2_2_0/`.

`CommonPathPrefixKeyGenerator` tokenizes the path by `/` and searches for the longest prefix starting from the last token.
If no prefix is found, the local path is used as-is. 
If there is a prefix, the next token after the match (a subdirectory) is mapped to a generated name.
Thus, `CommonPathPrefixKeyGenerator` supports more than one level of pseudo-random names, it is required, however, that `createDirectory` for the parent path to be called before it is called for the subdirectories on any level.

kssenii replied:
> Currently, the disk does not support moveFile required for materialize projections
Also, from the task description (projection is mutation):

Ah, I did not know it needed `moveFile`. Could we explicitly disable projections then?

> Not related to the PR, it looks like there is a bug in change setting that became visible if the disk is specified through SETTINGS disk = disk(...)in CREATE TABLE and then the table is altered

Wow, thanks for catching. Created an issue https://github.com/ClickHouse/ClickHouse/issues/63019, I'll fix it (it was me who implemented this feature to specify disk via `SETTINGS disk = disk(...)`)

> ALTER TABLE MOVE PARTITION TO TABLE requires hard link creation

ah, ok, I thought it does `moveFile` instead, then yes - no need to support this

> I changed it because prefix.path will be created for directories that are static (will never be be renamed or moved), such as the top-level subdirectories store, 7d5 and 7d563340-6f73-463b-bc78-95baf347d19a (table uuid) in the example store/7d5/7d563340-6f73-463b-bc78-95baf347d19a/tmp_insert_all_2_2_0/.

Yeah, makes sense. Could you please add a comment about this in code? So that it will be clear for the code reader why we store some directories as is. 

btw may be we could get rid of `./store/` in object storage keys? or at least from `./` in the beginning? Getting rid of `store` could be good for s3 read performance, as it behaves slower when we read a lot of keys with the same prefix concurrently.

jkartseva replied:
> Could we explicitly disable projections then?

Explicitly disabled alters and mutations for plain type:
https://github.com/ClickHouse/ClickHouse/pull/61116/commits/1033ddfc9de311892ceb528dd0baebd96d53a985

kssenii commented on src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.cpp:
```diff
@@ -1,26 +1,90 @@
+    MetadataStorageFromPlainObjectStorage::PathMap result;
+
+    RelativePathsWithMetadata files;
+    object_storage->listObjects(root, files, 0);
```
It will be better to use `IObjectStorage::iterate` instead of `IObjectStorage::listObjects`, because:
1. it will be more memory efficient, the number of files we list can be really huge
2. we could parallelize loading of metadata on server startup,  `IObjectStorageIterator` is thread-safe. Let's add a setting in disk configuration like `load_metadata_threads` and use threadpool to load metadata?

jkartseva replied:
Is it tested with minio?
i'm getting
```
2024.04.26 02:28:04.184442 [ 357825 ] {} <Error> Application: Code: 499. DB::Exception: Could not list objects in bucket 'test' with prefix 'clickhouse-disks/', S3 exception: `InvalidArgument`, message: 'Unable to parse ExceptionName: InvalidArgument Message: The continuation token provided is incorrect'. (S3_ERROR), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x000000001349ac72
1. ./build/./src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000afa7939
2. ./contrib/llvm-project/libcxx/include/string:1499: DB::S3Exception::S3Exception<String, String, String, String>(Aws::S3::S3Errors, fmt::v8::basic_format_string<char, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type, fmt::v8::type_identity<String>::type>, String&&, String&&, String&&, String&&) @ 0x000000000ea32922
3. ./build/./src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp:147: DB::(anonymous namespace)::S3IteratorAsync::getBatchAndCheckNext(std::vector<DB::RelativePathWithMetadata, std::allocator<DB::RelativePathWithMetadata>>&) @ 0x000000000f3d2d7f
4. ./build/./src/Disks/ObjectStorages/ObjectStorageIteratorAsync.cpp:70: DB::IObjectStorageIteratorAsync::BatchAndHasNext std::__function::__policy_invoker<DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>::__call_impl<std::__function::__default_alloc_func<DB::IObjectStorageIteratorAsync::scheduleBatch()::$_0, DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>>(std::__function::__policy_storage const*) @ 0x000000000eaf2ae6
5. ./contrib/llvm-project/libcxx/include/__functional/function.h:0: ? @ 0x000000000eaef6d4
6. ./contrib/llvm-project/libcxx/include/future:1380: std::packaged_task<DB::IObjectStorageIteratorAsync::BatchAndHasNext ()>::operator()() @ 0x000000000eaefc56
7. ./base/base/../base/wide_integer_impl.h:817: ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000b0519cc
8. ./build/./src/Common/ThreadPool.cpp:0: ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() @ 0x000000000b0547f1
9. ./base/base/../base/wide_integer_impl.h:817: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000b04f002
10. ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000b0533ee
11. ? @ 0x00007dbff9494ac3
```

kssenii replied:
Ah, yes, this iterator is not used anywhere now, but it will be used for s3 table engine after https://github.com/ClickHouse/ClickHouse/pull/59767 is merged. It did not work there initially as well, but I made several fixes and it started to work. I will extract the fixes into a separate PR to faster merge them. If you want, you can copy them into your PR.

jkartseva replied:
I will leave it as-is for now, but follow up after either https://github.com/ClickHouse/ClickHouse/pull/59767 or the PR with essential fixes is merged.

jkartseva committed:
explicitly disallow ALTERs and mutations for plain

jkartseva commented on src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.h:
```diff
@@ -0,0 +1,80 @@
+        MetadataStorageFromPlainObjectStorage::PathMap & path_map_,
+        ObjectStoragePtr object_storage_);
+
+    void execute(std::unique_lock<SharedMutex> & metadata_lock) override;
+    void undo() override;
```
I agree the lock reference should be part of the `undo` signature.
I was reluctant to change it since that will also require the change of multiple `MetadataStorageFromDiskTransactionOperations` after I extracted the common interface: https://github.com/ClickHouse/ClickHouse/blob/master/src/Disks/ObjectStorages/MetadataStorageFromDiskTransactionOperations.h

There is no bug because `undo` is done [under the lock](https://github.com/ClickHouse/ClickHouse/pull/61116/files#diff-2b99730b3a5ca3f8a55c53fc2789914c93c90da4a7fed49fc2a604d8dd3e3232R60)
Since the current implementation is confusing I'll change it, should be a purely cosmetic change.

jkartseva committed:
Update src/Disks/ObjectStorages/CommonPathPrefixKeyGenerator.h

Co-authored-by: Kseniia Sumarokova <54203879+kssenii@users.noreply.github.com>

jkartseva committed:
Update src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

Co-authored-by: Kseniia Sumarokova <54203879+kssenii@users.noreply.github.com>

jkartseva committed:
address feedback - pt.1

jkartseva commented on tests/integration/test_s3_plain_rewritable/test.py:
```diff
@@ -0,0 +1,78 @@
+        )
+
+
+def test_restart():
```
I've used `pytest.mark.order` for ordering.

jkartseva committed:
address feedback - pt.2

jkartseva committed:
address feedback - pt.3

non-functional changes

jkartseva committed:
address feedback - pt.4

jkartseva commented on src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp:
```diff
@@ -0,0 +1,188 @@
+    if (write_finalized)
+        path_map.emplace(path_from, path_map.extract(path_to).mapped());
+
+    if (write_created)
```
Rollback should be identical for `write_finalized` and `write_created` code paths since `prefix.path` is overridden, not created anew.

jkartseva committed:
extract plain_rewritable metadata type

Make PlainRewritableObjectStorage generic;

Support config type
```
<s3_plain_rewritable>
    <type>object_storage</type>
    <object_storage_type>s3</object_storage_type>
    <metadata_type>plain_rewritable</metadata_type>
    <endpoint>https://s3.eu-west-1.amazonaws.com/clickhouse-eu-west-1.clickhouse.com/data/</endpoint>
    <use_environment_credentials>1</use_environment_credentials>
</s3_plain_rewritable>
```

jkartseva committed:
documentation for unbundled configuration

jkartseva commented on docs/en/operations/storing-data.md:
```diff
@@ -341,6 +341,23 @@ Configuration:
+Unlike `s3_plain` disk type, `s3_plain_rewritable` allows executing merges and supports INSERT operations.
+[Mutations](/docs/en/sql-reference/statements/alter#mutations) and replication of tables are not supported.
+
+A use case for this disk type are non-replicated `MergeTree` tables. e.g., system tables.
```
Added that to docs.

jkartseva commented on docs/en/operations/storing-data.md:
```diff
@@ -341,6 +341,23 @@ Configuration:
+    <type>s3_plain_rewritable</type>
+    <endpoint>https://s3.eu-west-1.amazonaws.com/clickhouse-eu-west-1.clickhouse.com/data/</endpoint>
+    <use_environment_credentials>1</use_environment_credentials>
+</s3_plain_rewritable>
```
It works now.

jkartseva committed:
use ordered map for path map

jkartseva commented on tests/queries/0_stateless/03008_s3_plain_rewritable.sql:
```diff
@@ -0,0 +1,22 @@
+
+drop table if exists test_mt;
+create table test_mt (a Int32, b Int64) engine = MergeTree() order by a
+settings disk = disk(
```
Could you please point me to where it is located?

kssenii replied:
_log system tables defined for CI are in `ClickHouse/tests/config/config.d/`, for example, https://github.com/ClickHouse/ClickHouse/blob/master/tests/config/config.d/blob_storage_log.xml. It is possible to define `engine` parameter in this config, where we can also add `storage_policy`, example: https://github.com/ClickHouse/ClickHouse/blob/5139ef7b9109a9d4a3822c88f77fca457ef41887/tests/integration/test_system_logs/configs/config.d/system_logs_engine.xml#L3
We could add `s3_plain_rewritable` for `blob_storage_log` from the example above. Possibly it will be better to do this only for a tests run which runs all tests on top of s3, this can be checked here https://github.com/ClickHouse/ClickHouse/blob/5139ef7b9109a9d4a3822c88f77fca457ef41887/tests/config/install.sh#L159. E.g. in this `if` branch we could substitute blob_storage_log.xml to a different one (with a different storage policy) instead of what is added here https://github.com/ClickHouse/ClickHouse/blob/5139ef7b9109a9d4a3822c88f77fca457ef41887/tests/config/install.sh#L22

jkartseva committed:
remove path normalization

kssenii commented on src/Disks/IDisk.h:
```diff
@@ -363,6 +363,9 @@ class IDisk : public Space
 
     virtual bool isWriteOnce() const { return false; }
 
+    /// Whether this disk support mutations.
+    virtual bool isMutable() const { return true; }
```
May be better to explicitly name it `supportsHardlinks`? Because `isMutable` is kind of the opposite of `IDisk::isReadOnly`, so might be confusing

jkartseva committed:
method rename

jkartseva commented:
Getting an error when trying to specify ```<engine>Engine = MergeTree PARTITION BY ... </engine>```

```
2024.04.30 08:11:13.067385 [ 771 ] {} <Error> Application: Caught exception while loading metadata: Code: 36. DB::Exception: If 'engine' is specified for system table, PARTITION BY parameters should be specified directly inside 'engine' and 'partition_by' setting doesn't make sense. (BAD_ARGUMENTS), Stack trace (when copying this message, always include the lines below):
```
I'll remove the commit with `blog_storage_log` config from the stack and follow-up separately.
